rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+`https://github.com/flavio/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2` | ✅ | The policy is coming from `https://github.com/flavio/`
+`https://github.com/kubewarden/privileged-pods/.github/workflows/release.yml@refs/tags/v0.2.0` | ✅ | The policy is coming from `https://github.com/kubewarden/`
+`https://github.com/kubewarden/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.3.0` | ✅ | The policy is coming from `https://github.com/kubewarden/`
+`https://github.com/alice/fancy-policy/.github/workflows/cd.yml@refs/heads/main` | ✅ | The policy is coming from `https://github.com/alice/fancy-policy/`
+`https://github.com/alice/another-policy/.github/workflows/cd.yml@refs/heads/main` | ❌ | The policy is coming from a repository owned by Alice, but not trusted by the configuration
+
+Given the previous configuration, we would also trust all the policies that have been
+signed with the `john@acme.org` email, as long as the identity provider is the
+GitHub website (this is the identity provider used by "human-driven" keyless signatures
+relying on the GitHub identity provider).
+
+**Note well:** `issuers` defines a per-issuer allow list. That means the policies
+**do not have** to be signed by all the subjects mentioned inside of the list.
+They must all be signed by at least one of these issuers.
+
+### Annotations
+
+This is an entirely optional section that applies both to key and keyless
+signatures.
+
+Cosign allows the signer to specify custom key-value attributes to embedded
+into the signature.
+
+```yaml
+annotations:
+  env: prod
+  foo: bar
+```
+
+Given the previous configuration, all the policies must include an annotation
+with key `env` and value `prod`, plus another one with key `foo` and value `bar`.
+
+## Configuration scenarios
+
+As stated above the verification settings are made by three sections:
+
+  * `public_keys`
+  * `issuers`
+  * `annotations`
+
+The following table describes how verification happens when these fields
+are specified (✅) or left empty (❌).
+
+| `public_keys` | `issuers` | annotations | Behaviour |
+|-------------|-----------|-------------|-----------|
+| ✅ | ❌ | ❌ | policies must be signed by at least <threshold> keys |
+| ✅ | ✅ | ❌ | policies must be signed by at least <threshold> keys, **plus** they must be signed by one of the issuers |
+| ❌ | ✅ | ❌ | policies must be signed by one of the issuers |
+| ❌ | ✅ | ✅ | policies must be signed by one of the issuers, **plus** they must satisfy all the annotations |
+| ✅ | ❌ | ✅ | policies must be signed by at least <threshold> keys, **plus** they must satisfy all the annotations |
+| ✅ | ✅ | ✅ | policies must be signed by at least <threshold> keys, plus they must be signed by one of the issuers, **plus** they must satisfy all the annotations |
+| ❌ | ❌ | ✅ | this configuration is wrong, it should result in an error |
```

Comment:
With the `anyOf` approach, the last configuration would be possible (just, checking that there's any signature that marks the images as `prod` for example, regardless of the signer).

Reply:
With `anyOf`, `allOf`, we may just drop this section.

Reply:
I am also in favour of the `anyOf` and `allOf` approach in principle. I guess they are exclusive with each other, right? In that you either provide `anyOf` or `allOf`, and beneath the one that has been chosen the list of signatures. Right?

Reply:
They aren't exclusive, you could provide both. And one can list `pub_keys`, `keyless tuples`, or `annotations` on both of them.

Reply:
At first it looked a bit confusing, but if we make it clear that if both are provided then both should be met I think we are fine.

Reply:
Now obsolete, because of the changes done to `verification.yml` format